enhancements/20230907-snapshot-space-management.md

```diff
@@ -0,0 +1,94 @@
+# Snapshot Space Management
+
+## Summary
+
+This proposal leverages [LH Snapshot Space Management](https://github.com/longhorn/longhorn/issues/6563) to control snapshot space usage in the system. LH introduces two new attributes `snapshotMaxCount` and `snapshotMaxSize` in Volume. Harvester can use these parameters to set upper bound for snapshot space usage.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/4478
+
+## Motivation
+
+### Goals
+
+- Set snapshot count and size limitations for each volume.
+
+## Proposal
+
+Use a PVC controller to watch PVCs. If there is `harvesterhci.io/snapshotMaxCount` or `harvesterhci.io/snapshotMaxSize` annotation on a PVC, the system uses the value to update the related LH Volume.
+
+### User Stories
+
+With snapshot space management, snapshot space usage is controllable. Users can set snapshot limitation and evaluate maximum space usage.
+
+### User Experience In Detail
+
+Before snapshot space management:
+The default maximum snapshot count for each volume is `250`. This is a constants value in the [source code](https://github.com/longhorn/longhorn-engine/blob/8b4c80ab174b4f454a992ff998b6cb1041faf63d/pkg/replica/replica.go#L33) and users don't have a way to control it. If a volume size is 1G, the maximum snapshot space usage will be 250G.
+
+After snapshot space management:
+There is a configurable default maximum snapshot count setting. Users can update it to overwrite the fixed value in the system.
+Users can set different maximum snapshot count and size on each volume. A more important volume can have more snapshot space usage.
+
+### API changes
+
+No harvester API changes are needed. The changes will only be in the pvc controller logic.
+
+## Design
+
+### Implementation Overview
+
+#### Settings
+
+Add a harvester setting `snapshot-max-count`. The range of the value is same as LH setting `snapshot-max-count`. It should be between `2` to `250`. The [Setting controller](https://github.com/harvester/harvester/blob/ea814927a14f99b0f0ef9fe74c37dbea3e6e95df/pkg/controller/master/setting/register.go#L66-L80) will overwrite the value to LH setting.
```

Comment:
@innobead Do we know what's the max number of system snapshots? Is there a limitation? Thanks!